[bug] creating iOS packages using simple toolchains broken as of conan 1.26

Consider this simple iOS toolchain (based on [this toolchain](https://github.com/microblink/build/blob/master/toolchains/ios.toolchain.cmake)):

```cmake
cmake_minimum_required( VERSION 3.14 )

set( CMAKE_C_COMPILER_ID   AppleClang )
set( CMAKE_CXX_COMPILER_ID AppleClang )

if( NOT ${CMAKE_GENERATOR} MATCHES "Xcode" )
    message( FATAL_ERROR "iOS toolchain supports only XCode generator" )
endif()

# Standard settings
set( CMAKE_SYSTEM_NAME      iOS    )
set( CMAKE_OSX_SYSROOT      "iphoneos" )
set( CPACK_SYSTEM_NAME      iOS    )
set( CMAKE_SYSTEM_VERSION   8.0    )
set( CMAKE_SYSTEM_PROCESSOR arm    )
set( CMAKE_OSX_ARCHITECTURES armv7 armv7s arm64 i386 x86_64 )

# Skip the platform compiler checks for cross compiling (or not)...
set( CMAKE_CXX_COMPILER_WORKS true CACHE STRING "Skip CMake compiler detection (requires a functioning code signing identity and provisioning profile)." )
set( CMAKE_C_COMPILER_WORKS   ${CMAKE_CXX_COMPILER_WORKS} )
set( CMAKE_XCODE_EFFECTIVE_PLATFORMS "-universal;-iphonesimulator;-iphoneos;" )
set( CMAKE_IOS_DEVELOPER_ROOT        "/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer" )
set( CMAKE_XCODE_ATTRIBUTE_SDKROOT   iphoneos ) # "Latest iOS"

set( XCODE_ATTRIBUTE_CFLAGS_armv7  "-mcpu=cortex-a9 -mfpu=neon" ) 
set( XCODE_ATTRIBUTE_CFLAGS_armv7s "-mcpu=swift"                )
set( CMAKE_XCODE_ATTRIBUTE_OTHER_CFLAGS[arch=armv7]          "${XCODE_ATTRIBUTE_CFLAGS_armv7}  $(OTHER_CFLAGS)"         )
set( CMAKE_XCODE_ATTRIBUTE_OTHER_CFLAGS[arch=armv7s]         "${XCODE_ATTRIBUTE_CFLAGS_armv7s} $(OTHER_CFLAGS)"         )
set( CMAKE_XCODE_ATTRIBUTE_OTHER_CPLUSPLUSFLAGS[arch=armv7]  "${XCODE_ATTRIBUTE_CFLAGS_armv7}  $(OTHER_CPLUSPLUSFLAGS)" )
set( CMAKE_XCODE_ATTRIBUTE_OTHER_CPLUSPLUSFLAGS[arch=armv7s] "${XCODE_ATTRIBUTE_CFLAGS_armv7s} $(OTHER_CPLUSPLUSFLAGS)" )
```

Let's create a simple package that contains this toolchain:

```python
import os

from conans import ConanFile


class IostoolchainConan(ConanFile):
    name = "IosToolchain"
    version = "1.2.0"
    # obtain sources, etc.

    def package(self):
        self.copy('ios.toolchain.cmake')

    def package_info(self):
        self.env_info.CONAN_CMAKE_TOOLCHAIN_FILE = os.path.join(self.package_folder, 'ios.toolchain.cmake')
        self.env_info.CONAN_CMAKE_CUSTOM_iOS_ALLOW_UNIVERSAL_BUILD = 'ON'

    def package_id(self):
        self.info.header_only()
```

Also, let's create a conan profile using this toolchain: 

```
[settings]
os=iOS
os.version=8.0
os_build=Macos
arch=ios_fat
arch_build=x86_64
compiler=apple-clang
compiler.libcxx=libc++
build_type=Release
compiler.version=11.0.3

[build_requires]
IosToolchain/1.2.0@microblink/stable

[env]
CONAN_CMAKE_GENERATOR = Xcode
```

Now, let's use this profile to create some package

```
conan create . user/testing -pr ios-profile
```

With Conan 1.25.2 and earlier, everything works. With Conan 1.26.1 and later (also tested with 1.28.0), I get the following CMake configuration error:

```
CMake Error at /Applications/CMake.app/Contents/share/cmake-3.18/Modules/Platform/iOS-Initialize.cmake:4 (message):
  
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk
  is not an iOS SDK
...
-- Configuring incomplete, errors occurred!
```

The problem can be traced to how conan's CMake build helper invokes cmake: 

```
cmake -G "Xcode" -DCMAKE_OSX_ARCHITECTURES="None" -DCMAKE_OSX_SYSROOT="/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk" -DCMAKE_TOOLCHAIN_FILE="/Users/dodo/.conan/data/IosToolchain/1.2.0/microblink/stable/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/ios.toolchain.cmake" -DCONAN_IN_LOCAL_CACHE="ON" -DCONAN_COMPILER="apple-clang" -DCONAN_COMPILER_VERSION="11.0.3" -DCONAN_LIBCXX="libc++" ...
```

The problematic part is `-DCMAKE_OSX_SYSROOT="/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk"` - this overrides the `CMAKE_OSX_SYSROOT` variable as set by the toolchain. Also problematic is `-DCMAKE_OSX_ARCHITECTURES="None"` that does the same.

When executed with `CONAN_PRINT_RUN_COMMANDS=1` by using conan 1.25.2 (or earlier), it can be seen that those two variables are not given via command-line prior to `CMAKE_TOOLCHAIN_FILE`.

I know that from v1.26 there is support for new toolchain mechanism, but I didn't get how should I modify my toolchain files to work with the new conan.

--- a/conans/client/build/cmake_flags.py
+++ b/conans/client/build/cmake_flags.py
@@ -295,7 +295,7 @@ class CMakeDefinitionsBuilder(object):
                                                  self._generator, self._output))
 
         # don't attempt to override variables set within toolchain
-        if tools.is_apple_os(os_) and "CONAN_CMAKE_TOOLCHAIN_FILE" not in os.environ:
+        if tools.is_apple_os(os_) and "CONAN_CMAKE_TOOLCHAIN_FILE" not in os.environ and "CMAKE_TOOLCHAIN_FILE" not in definitions:
             apple_arch = tools.to_apple_arch(arch)
             if apple_arch:
                 definitions["CMAKE_OSX_ARCHITECTURES"] = apple_arch

--- a/conans/client/build/cmake_flags.py
+++ b/conans/client/build/cmake_flags.py
@@ -294,12 +294,18 @@ class CMakeDefinitionsBuilder(object):
         definitions.update(build_type_definition(self._forced_build_type, build_type,
                                                  self._generator, self._output))
 
-        if tools.is_apple_os(os_):
-            definitions["CMAKE_OSX_ARCHITECTURES"] = tools.to_apple_arch(arch)
+        # don't attempt to override variables set within toolchain
+        if tools.is_apple_os(os_) and "CONAN_CMAKE_TOOLCHAIN_FILE" not in os.environ:
+            apple_arch = tools.to_apple_arch(arch)
+            if apple_arch:
+                definitions["CMAKE_OSX_ARCHITECTURES"] = apple_arch
             # xcrun is only available on macOS, otherwise it's cross-compiling and it needs to be
-            # set within CMake toolchain
-            if platform.system() == "Darwin":
-                definitions["CMAKE_OSX_SYSROOT"] = tools.XCRun(self._conanfile.settings).sdk_path
+            # set within CMake toolchain. also, if SDKROOT is set, CMake will use it, and it's not
+            # needed to run xcrun.
+            if platform.system() == "Darwin" and "SDKROOT" not in os.environ:
+                sdk_path = tools.XCRun(self._conanfile.settings).sdk_path
+                if sdk_path:
+                    definitions["CMAKE_OSX_SYSROOT"] = sdk_path
 
         definitions.update(self._cmake_cross_build_defines())
         definitions.update(self._get_cpp_standard_vars())

--- a/conans/client/tools/apple.py
+++ b/conans/client/tools/apple.py
@@ -32,13 +32,11 @@ def apple_sdk_name(settings):
                 'iOS': 'iphonesimulator',
                 'watchOS': 'watchsimulator',
                 'tvOS': 'appletvsimulator'}.get(str(os_))
-    elif str(arch).startswith('arm'):
-        return {'iOS': 'iphoneos',
-                'watchOS': 'watchos',
-                'tvOS': 'appletvos'}.get(str(os_))
     else:
-        return None
-
+        return {'Macos': 'macosx',
+                'iOS': 'iphoneos',
+                'watchOS': 'watchos',
+                'tvOS': 'appletvos'}.get(str(os_), None)
 
 def apple_deployment_target_env(os_, os_version):
     """environment variable name which controls deployment target"""

--- a/conans/test/functional/build_helpers/cmake_apple_test.py
+++ b/conans/test/functional/build_helpers/cmake_apple_test.py
@@ -1,0 +1,59 @@
+import platform
+import unittest
+from parameterized import parameterized
+
+from conans.client.build.cmake import CMake
+from conans.client.conf import get_default_settings_yml
+from conans.model.settings import Settings
+from conans.test.utils.mocks import MockSettings, ConanFileMock
+
+
+@unittest.skipUnless(platform.system() == "Darwin", "Only for MacOS")
+class CMakeAppleTest(unittest.TestCase):
+    @parameterized.expand([('x86', 'Macos', 'i386', 'MacOSX.platform'),
+                           ('x86_64', 'Macos', 'x86_64', 'MacOSX.platform'),
+                           ('armv7', 'Macos', 'armv7', 'MacOSX.platform'),
+                           ('armv8', 'Macos', 'arm64', 'MacOSX.platform'),
+                           ('x86', 'iOS', 'i386', 'iPhoneSimulator.platform'),
+                           ('x86_64', 'iOS', 'x86_64', 'iPhoneSimulator.platform'),
+                           ('armv7', 'iOS', 'armv7', 'iPhoneOS.platform'),
+                           ('armv8', 'iOS', 'arm64', 'iPhoneOS.platform'),
+                           ('x86', 'watchOS', 'i386', 'WatchSimulator.platform'),
+                           ('x86_64', 'watchOS', 'x86_64', 'WatchSimulator.platform'),
+                           ('armv7', 'watchOS', 'armv7', 'WatchOS.platform'),
+                           ('armv8', 'watchOS', 'arm64', 'WatchOS.platform'),
+                           ('x86', 'tvOS', 'i386', 'AppleTVSimulator.platform'),
+                           ('x86_64', 'tvOS', 'x86_64', 'AppleTVSimulator.platform'),
+                           ('armv7', 'tvOS', 'armv7', 'AppleTVOS.platform'),
+                           ('armv8', 'tvOS', 'arm64', 'AppleTVOS.platform')
+                           ])
+    def test_cmake_definitions(self, conan_arch, conan_os, expected_arch, expected_os):
+        settings = Settings.loads(get_default_settings_yml())
+        settings.os = conan_os
+        settings.compiler = "apple-clang"
+        settings.compiler.version = "11.0"
+        settings.arch = conan_arch
+
+        conanfile = ConanFileMock()
+        conanfile.settings = settings
+        cmake = CMake(conanfile)
+
+        self.assertEqual(cmake.definitions["CMAKE_OSX_ARCHITECTURES"], expected_arch)
+        self.assertIn(expected_os, cmake.definitions["CMAKE_OSX_SYSROOT"])
+
+    @parameterized.expand([('iOS', 'iPhoneOS.platform'),
+                           ('Macos', 'MacOSX.platform'),
+                           ('watchOS', 'WatchOS.platform'),
+                           ('tvOS', 'AppleTVOS.platform')
+                           ])
+    def test_custom_settings(self, conan_os, expected_os):
+        settings = MockSettings({"os": conan_os,
+                                 "compiler": "apple-clang",
+                                 "compiler.version": "11.0",
+                                 "arch": "ios_fat"})
+        conanfile = ConanFileMock()
+        conanfile.settings = settings
+        cmake = CMake(conanfile)
+
+        self.assertNotIn("CMAKE_OSX_ARCHITECTURES", cmake.definitions)
+        self.assertIn(expected_os, cmake.definitions["CMAKE_OSX_SYSROOT"])

--- a/conans/test/unittests/util/apple_test.py
+++ b/conans/test/unittests/util/apple_test.py
@@ -42,7 +42,6 @@ class AppleTest(unittest.TestCase):
         self.assertIsNone(tools.to_apple_arch('mips'))
 
     def test_apple_sdk_name(self):
-
         self.assertEqual(tools.apple_sdk_name(FakeSettings('Macos', 'x86')), 'macosx')
         self.assertEqual(tools.apple_sdk_name(FakeSettings('Macos', 'x86_64')), 'macosx')
         self.assertEqual(tools.apple_sdk_name(FakeSettings('iOS', 'x86_64')), 'iphonesimulator')
@@ -52,6 +51,13 @@ class AppleTest(unittest.TestCase):
         self.assertEqual(tools.apple_sdk_name(FakeSettings('tvOS', 'x86')), 'appletvsimulator')
         self.assertEqual(tools.apple_sdk_name(FakeSettings('tvOS', 'armv8')), 'appletvos')
         self.assertIsNone(tools.apple_sdk_name(FakeSettings('Windows', 'x86')))
+
+    def test_apple_sdk_name_custom_settings(self):
+        self.assertEqual(tools.apple_sdk_name(FakeSettings('Macos', 'ios_fat')), 'macosx')
+        self.assertEqual(tools.apple_sdk_name(FakeSettings('iOS', 'ios_fat')), 'iphoneos')
+        self.assertEqual(tools.apple_sdk_name(FakeSettings('watchOS', 'ios_fat')), 'watchos')
+        self.assertEqual(tools.apple_sdk_name(FakeSettings('tvOS', 'ios_fat')), 'appletvos')
+        self.assertIsNone(tools.apple_sdk_name(FakeSettings('ConanOS', 'ios_fat')))
 
     def test_deployment_target_env_name(self):
         self.assertEqual(tools.apple_deployment_target_env('Macos', "10.1"),

